File: app/src/main/java/com/warchaser/musicplayer/tools/CallObserver.java
package com.warchaser.musicplayer.tools;

import android.content.Intent;

import java.util.ArrayList;

/**
 * Created by Wucn on 2017/1/26.
 * 观察者类
 * 双向通知
 */
public class CallObserver {
    private static ArrayList<UIObserver> mObservers = new ArrayList<UIObserver>();

    private CallObserver() {

    }

    public static void registerObserver(UIObserver observer) {
        if(mObservers == null){
            return;
        }

        mObservers.add(observer);
    }

    /**
     * 通知UI刷新
     * */
    public synchronized static void callObserver(final Intent intent) {
        process((observer) -> observer.notifySeekBar2Update(intent));
    }

    /**
     * 通知UI更新逻辑
     *
     * @param repeatTime 1 重复一次(playOrPause); 2重复两次(next)
     * @return boolean 是否存在UI Observer
     */
    public synchronized static boolean callPlay(final int repeatTime) {
        final ValueBean bean = new ValueBean();

        process(observer -> {
            observer.notify2Play(repeatTime);
            bean.setFlag(true);
        });

        return !bean.isFlag();
    }

    public static void removeAllObservers() {
        if (mObservers != null) {
            mObservers.clear();
            mObservers = null;
        }
    }

    public static void removeSingleObserver(UIObserver observer) {
        if (mObservers != null) {
            mObservers.remove(observer);
        }
    }

    public static boolean isNeedCallObserver() {
        return mObservers != null && !mObservers.isEmpty();
    }

    public static void stopUI() {
        processAll(UIObserver::stopServiceAndExit);
        removeAllObservers();
    }

    /**
     * 根据条件执行
     * */
    private static void process(SubFunction<UIObserver> observerSubFunction) {
        if (!isNeedCallObserver()) {
            return;
        }

        int size = size();
        for (int i = 0; i < size; i++) {
            UIObserver observer = mObservers.get(i);
            if (observer != null && observer.getObserverEnabled()) {
                observerSubFunction.processor(observer);
            }
        }
    }

    /**
     * 无条件执行
     * */
    private static void processAll(SubFunction<UIObserver> observerSubFunction) {
        if (!isNeedCallObserver()) {
            return;
        }

        int size = size();
        for (int i = 0; i < size; i++) {
            UIObserver observer = mObservers.get(i);
            if (observer != null) {
                observerSubFunction.processor(observer);
            }
        }
    }

    private static int size(){
        return mObservers.size();
    }

    private static class ValueBean {
        private boolean isFlag = false;

        public boolean isFlag() {
            return isFlag;
        }

        public void setFlag(boolean flag) {
            isFlag = flag;
        }
    }

    private interface SubFunction<T> {
        void processor(T t);
    }

}

File: app/src/main/java/com/warchaser/musicplayer/tools/CallObserver.kt
package com.warchaser.musicplayer.tools

import android.content.Intent
import java.util.ArrayList

/**
 * Created by Achan on 2020/10/26
 * 观察者类
 * 双向通知
 * */

private typealias Processor<T> = (t : T) -> Unit

class CallObserver private constructor(){

    companion object{
        @JvmStatic
        val instance = Holder.mHolder
        private val mObservers : ArrayList<UIObserver?> = ArrayList()

        private class ValueBean{
            var isFlag : Boolean = false
        }
    }

    private object Holder{
        val mHolder = CallObserver()
    }

    fun registerObserver(observer: UIObserver){
        mObservers.add(observer)
    }

    fun isNeedCallObserver() : Boolean{
        return mObservers.isNotEmpty()
    }

    private fun size() : Int{
        return mObservers.size
    }

    fun removeSingleObserver(observer: UIObserver){
        mObservers.remove(observer)
    }

    fun removeAllObservers(){
        mObservers.clear()
    }

    /**
     * 根据条件执行
     * */
    private fun process(observerSubFunction : SubFunction<UIObserver>){
        if(!isNeedCallObserver()){
            return
        }

        mObservers.forEach {
            it?.run {
                if(observerEnabled){
                    observerSubFunction.processor(this)
                }
            }
        }
    }

    /**
     * 无条件执行
     * */
    private fun processAll(observerSubFunction : SubFunction<UIObserver>){
        if(!isNeedCallObserver()){
            return
        }

        mObservers.forEach {
            it?.run {
                observerSubFunction.processor(this)
            }
        }
    }

    /**
     * 通知UI更新逻辑
     *
     * @param repeatTime 1 重复一次(playOrPause); 2重复两次(next)
     * @return boolean 是否存在UI Observer
     */
    fun callPlay(repeatTime : Int) : Boolean{
        val bean = ValueBean()
        process(registerFunction {
            processor {
                it.notify2Play(repeatTime)
                bean.isFlag = true
            }
        })
        return !bean.isFlag
    }

    /**
     * 通知UI刷新
     * */
    @Synchronized
    fun callObserver(intent: Intent){
        process(registerFunction {
            processor { it.notifySeekBar2Update(intent) }
        })
    }

    fun stopUI(){
        processAll(registerFunction {
            processor { it.stopServiceAndExit() }
        })
        removeAllObservers()
    }

    private interface SubFunction<T>{
        fun processor(t : T)
    }

    inner class SubFunctionDSL<T : UIObserver> : SubFunction<T>{

        private var mProcessor : Processor<T>? = null

        override fun processor(t: T) {
            mProcessor?.invoke(t)
        }

        fun processor(processor: Processor<T>){
            mProcessor = processor
        }
    }

    private fun registerFunction(function : SubFunctionDSL<UIObserver>.() -> Unit) = SubFunctionDSL<UIObserver>().also(function)

}